python/cuml/dask/common/func.py

```diff
@@ -0,0 +1,61 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import dask
+from toolz import first
+
+
+@dask.delayed
+def reduce_func_add(a, b): return a + b if b is not None else a
+
+
+def tree_reduce(delayed_objs, delayed_func=reduce_func_add):
+    """
+    Performs a binary tree reduce on an associative
+    and commutative function in parallel across
+    Dask workers.
+
+    TODO: investigate methods for doing intra-node
+    before inter-node reductions.
+
+    Parameters
+    ----------
+    delayed_func : dask.delayed function
+        Delayed function to use for reduction. The reduction function
+        should be able to handle the case where the second argument is
+        None, and should just return a in this case. This is done to
+        save memory, rather than having to build an initializer for
+        the starting case.
+    delayed_objs : array-like of dask.delayed
+        Delayed objects to reduce
+
+    Returns
+    -------
+    reduced_result : dask.delayed
+        Delayed object containing the reduced result.
+    """
+    while len(delayed_objs) > 1:
```

Comment:
If `len(delayed_objs)` is `1` in the beginning, the while loop won't be entered and the result will not be a delayed object. This causes inconsistent return type. 
- `tree_reduce([1, 2])  # returns a delayed object`
- `tree_reduce([1])  # returns 1`

I suggest returning a delayed object if `len(delayed_objs) == 1`, but we could also call `.compute` on the result after the while loop if it was entered.
```suggestion
     if len(delayed_objs) == 1:
         return dask.delayed(first(delayed_objs))
     while len(delayed_objs) > 1:
```

Reply:
Also if `len(delayed_objs)` is `0`, the call to `first(delayed_objs)` will raise an exception. Do we want to return `dask.delayed(None)` in that case?

Reply:
The 0 case is interesting because I don't see many places where we'll be calling this on 0 partitions, though I can short circuit the return if you'd like. 

> This causes inconsistent return type.

This function takes a list of `dask.delayed objects` so `[1, 2]` is not a valid input. When valid input is provided, this will always return a delayed object.